ds701_book/20-Recommender-Systems.qmd

```diff
@@ -303,57 +360,126 @@ We used __$k$-nearest neighbors__.   Here $k$ = 2.
 
 For now, let's just say we determine the similarities as:
 
-$$ s_{13} = 0.2 $$
+$$
+s_{13} = 0.2 
+$$
 
-$$ s_{16} = 0.3 $$
+$$
+s_{16} = 0.3 
+$$
 
-<center>
-    
-<img src="figs/L20-u-u-cf-3.png" alt="Figure" width="60%">
-    
-</center>
+:::
+::::
+
+---
 
 These similarity scores tell us how much weight to put on the rating of the other items.
 
-So we can form a prediction of $\hat{r}_{15}$ as:
-    
-$$ \hat{r}_{15} = \frac{s_{13} \cdot r_{35} + s_{16} \cdot r_{65}}{s_{13} + s_{16}} = \frac{0.2 \cdot 2 + 0.3 \cdot 3}{0.2 + 0.3} = 2.6 $$
+![](figs/L20-u-u-cf-4.png){width="50%" fig-align="center"}
 
-<center>
-    
-<img src="figs/L20-u-u-cf-4.png" alt="Figure" width="60%">
-    
-</center>
+So we can form a prediction of $\hat{r}_{15}$ as:
+  
+$$
+\hat{r}_{15} = \frac{s_{13} \cdot r_{35} + s_{16} \cdot r_{65}}{s_{13} + s_{16}} = \frac{0.2 \cdot 2 + 0.3 \cdot 3}{0.2 + 0.3} = 2.6 
+$$
 
-### Similarity
+## Similarity
 
 How should we assess similarity of items?
 
-A reasonable approach is to consider items similar if their ratings are __correlated.__
+A reasonable approach is to consider items similar if their ratings are
+__correlated__, for example using the Pearson correlation coefficient
+$r$[^pearson].
 
-So we can use the Pearson correlation coefficient $r$.
+[^pearson]: See this part of the Course Notes for the definition of Pearson correlation coefficient.
 
 However, note that two items will not have ratings in the same positions.
 
-<center>
-    
-<img src="figs/L20-corr-support.png" alt="Figure" width="60%">
-    
-</center>
+![](figs/L20-corr-support.png){width="60%" fig-align="center"}
 
 So we want to compute correlation only over the users who rated both the items.
 
-In some cases we will need to work with binary $r_{ui}$s.  
+## Example
+
+::: {.content-visible when-profile="slides"}
+![](figs/L20-corr-support.png){width="60%" fig-align="center"}
+:::
+
+Let's compute the Pearson correlation coefficient for the following two items:
+
+```{python}
+import numpy as np
+from IPython.display import display, Markdown
+
+ratings_item_i = [1, np.nan, np.nan, 5, 5, 3, np.nan, np.nan, np.nan, 4, 2, np.nan, np.nan, np.nan, np.nan, 4, np.nan, 5, 4, 1, np.nan]
+ratings_item_j = [np.nan, np.nan, 4, 2, 5, np.nan, np.nan, 1, 2, 5, np.nan, np.nan, 2, np.nan, np.nan, 3, np.nan, np.nan, np.nan, 5, 4]
+
+display(Markdown(f'Ratings for item $i$:\n{ratings_item_i}'))
```

Comment:
Can we display the lists on their own line? I think the wraparound makes it a bit difficult to see (but perhaps this depends on the screen/slide size).

Reply:
Can you suggest a way?

Reply:
These \n\n worked for me.
display(Markdown(f'Ratings for item $i$:\n\n{ratings_item_i}'))

If it somehow doesn't render well with this then it is a minor issue and we can ignore it.

Reply:
added